### ISSUE TITLE
Overhaul object copying in GAP and HPC-GAP, get rid of COPYING flag

### ISSUE THREAD

**Issue by fingolfin** [gap-system/gap#1876]
tl;dr: it would be nice if GAP and HPC-GAP used the same code for copying objects, and if we could get rid of the `COPYING` TNum flag in GAP, but I am not sure this is possibly w/o slowing down GAP:

Right now, copying a GAP kernel object (including lists, records, etc.) for most kinds of objects involves temporarily modifying the object, to insert a forward pointer to the copy, in order to allow copying recursively nested objects (including cases were this leads to cycles).
I.e. to copy `a:=[[]]; b:=a[1]; b[1]:=a;` resp. equivalently `a:=[[~]];b:=a[1];`, we must be careful to copy both lists `a` and `b` exactly once (and don't run into an infinite loop / recursion). Hence we "remember" for each object whether it was already copied, and the location of the copy.

Since we don't generally reserve space for the forwarding pointer, most objects do this by temporarily changing their TNUM (from `x` to `x+COPYING`), which (a) marks the object as being copied, and (b) ensures that any code which incorrectly accesses an object while it is being copied will immediately run into an error. After that, the object can create a copied object and then overwrite part of itself with the pointer to the copy. Once copying is done, all copied objects have to "clean up", i.e. restore the overwritten parts; luckily we have a copy of the object around, so that's no problem.

This all works fine for GAP (although it means we at time have objects which exist in a "weird" state). But it completely breaks down in HPC-GAP: we would like to copy objects while only holding a "readonly" lock to them, so that other threads which also want to access the object(s) {e..g to copy them, too) don't have to wait for us. But if we only hold a read lock, we must not modify those objects, as another thread may be in the middle of reading from them!

The first solution for this is to simply take a write lock on the objects. This may be fine for some operations, but it can easily lead to very bad lock contention; it might also be problematic in light of deadlocks (in the above example, if thread 1 tries to copy object `a`, and thread 2 tries to copy object `b`, both at the same time, they'll run into the classic situation were each has a write lock one of the two objects it needs, while the other object is locked by another thread). While it's possible to resolve that via various heuristics, it's unfortunate.

Indeed, the original algorithm has the nice property that it has to touch ever object essentially only once. In HPC-GAP, we either have to iterate over them three times (once to acquire locks, a second time to make the copies; a third time to release the locks); or else we do all of them at once, but then we may run into a situation we can't proceed (i.e. encountering an object we can't get write access to), at which point we are forced to perform a roll back and discard the partial copy we made. This adds to the cost.

So, what else can we do? If we are not allowed to modify the objects to store a forwarding pointer, we can't even easily mark them (remember that multiple threads may be copying the same object, and so they all would have to get a different "bit" in the object to track that). 

One alternative then is to use an `OBJ_MAP` to track all objects we already copied, as well as their copies (i.e. we map the objects to their copies). We then still have some choices: e.g. we can either iterate over the objects three times (once to read lock them all; if that worked, to copy them, and finally to release the locks, although that can now also be done via the data stored in the `OBJ_MAP`);  or only once, but with the provision that we may have to discard a partial copy (but that is now of course much cheaper, as we did not modify the originals, and thus don't have to restore any changes inside of them.

This avoids lock contention as long as no thread tries to write to the objects being copied. The main drawback is that of course a lookup in an `OBJ_MAP` is quite a bit more expensive than simply extracting a forwarding pointer from directly inside an object. Though in practice, this may not matter much; unless you have a zillion of objects nested into each. 

Indeed, this is then also more or less the approach taken by `HPC-GAP`, via `CopyReachableObjectsFrom` from `src/traverse.c`. I tried to make that stand out a bit more via PR #1875

Now, this approach could certainly also be used in GAP (we would simply leave out the locking), but it may be slower, due to the overhead of looking up objects in a hash map. On the upside, we don't have to restore forwarding pointers, which is nice. So one question is: How is copying performance affected, exactly? Benchmarks would be useful. 

One advantage of using this code is that it makes copying more robust: Right now, if a copying function runs into an error, I think we don't even try to recover; we might just end up crashing and/or with corrupted data. The new approach should avoid all that "for free".

@rbehrends once mentioned some ideas to reduce the overhead of the object map, but I have no idea about how.

**Comment by fingolfin** [comment]
Progress report: I got rid of `COPYING` in PR #2777. Instead, there is now a single `T_COPYING` tnum. In non-HPC-GAP, objects being copied temporarily have their tnum changed to that. So this means there is no fundamental change to the discussion above. However, there is now a path visible to at least unify the functions that perform the actual copying.
